### PR TITLE
Introduce GoldenToolkitConfiguration.enableRealShadows

### LINKED ISSUE #72: Support for "real shadows"
Hi,

first, many thanks for this project, we started to use it this week, it's already paying off !

One of the things that we quickly noticed while adding coverage for the components library we share between applications was that shadows were rendering in a strange way, and appeared as "opaque" blocks. 

After a bit of digging, I understand that, "by design", when running widget tests, shadows are rendered as solid blocks  : 

> since the rendering of shadows is not guaranteed to be pixel-for-pixel identical from version to version (or even from run to run).

(see https://api.flutter.dev/flutter/painting/debugDisableShadows.html ).

I found this trick to temporarily "re-enable" real shadows for a given test : https://stackoverflow.com/a/64675795/474763 and indeed it does the trick.

To make it a bit easier to apply across our tests, I ended up creating a small helper function like 

```dart
/// by default, Widget tests use opaque shadows to avoid golden test failures
/// we do want those on some tests to have images that look closer to reality
/// see https://stackoverflow.com/a/64675795/474763
Future<void> runWithRealShadows(Future<void> Function() callback) async {
  try {
    debugDisableShadows = false;
    await callback();
  } finally {
    debugDisableShadows = true;
  }
}
```

and using it in pretty much all my `testGoldens` tests like so : 

```dart
void main() {
  testGoldens('Boxes Demo', (tester) async {
    await runWithRealShadows(() async {
      await tester.pumpWidgetBuilder(
        BoxesDemo(),
        wrapper: designSystemWrapper,
      );
      await screenMatchesGolden(
        tester,
        'boxes_demo',
      );
    });
  });
}
```

This is fine for now, and does the job, but I'm wondering if, since `testGoldens` functions are already a wrapper around `testWidgets` , there would be away to package this functionality as part of `golden_toolkit` , so that we would be able to : 
- enable "real shadows" once and for all in the whole project (by setting it one in `flutter_test_config.dart` )
- maybe exposing it on `GoldenToolkitConfiguration` so it can be applied temporarily through `GoldenToolkit.runWithConfiguration()`

So I think that would be a nice addition to this library. If you think so too, I'd be happy to try and come up with a PR with an initial proposal of what it could look like.

### PR DESCRIPTION
**Pull request by tsimbalar**
## PR overview

This PR introduces a new flag `enableRealShadows` on `GoldenToolkitConfiguration`. 
This flags defaults to `false`.

When set to `true`, the global variable [`debugDisableShadows`](https://api.flutter.dev/flutter/painting/debugDisableShadows.html)  is set to `false` for the scope of the test (i.e. inside the `testWidget` body).

This enables to : 
- set globally that we want to use "real" shadows in our goldens by declaring the following in `flutter_test_config.dart` : 

```dart
Future<void> main(FutureOr<void> Function() testMain) async {
  return GoldenToolkit.runWithConfiguration(
    () async {
      await testMain();
    },
    config: GoldenToolkitConfiguration(
      enableRealShadows: true,
    ),
  );
}
```

- to set it locally / temporarily if we want to : 

```dart
GoldenToolkit.runWithConfiguration(() => {
      testGoldens(
          'this is my test',
          (tester) async {
        await tester.pumpWidgetBuilder(WidgetWithShadows());
        await screenMatchesGolden(
            tester, 'my_golden');
      })
    },
    config: GoldenToolkit.configuration.copyWith(enableRealShadows: true),
  );
});
```

**Important** : for this to work, `GoldenToolkit.runWithConfiguration` must be outside of the `testGoldens` call. 

As seen in the test, it will be ignored if `GoldenToolkit.runWithConfiguration`  is _inside_ the `testGoldens()` block.

## Remaining work 
- [ ] update example project
- [ ] update documentation (I'd rather wait until code has been reviewed, in case things need to be changed, renamed etc)

👇  Below, the commit message : 
-----
By default, Widget tests render shadows as opaque/solid blocks. The goal is to make rendering more deterministic and avoid false negative when using Golden tests.

When using the generated goldens as "living documentation" though, we do want those goldens to look as close as possible to reality.

Now `enableRealShadows` allow to use "real shadows" in generated goldens. 

This can be done globally by adding a global flag in a `flutter_test_config.dart` file

```dart
Future<void> main(FutureOr<void> Function() testMain) async {
  return GoldenToolkit.runWithConfiguration(
    () async {
      await testMain();
    },
    config: GoldenToolkitConfiguration(
      enableRealShadows: true,
    ),
  );
}
```

or closer to the test (note that `GoldenToolkit.runWithConfiguration` must surround the `testGoldens()` call for it to work)

```dart
GoldenToolkit.runWithConfiguration(() => {
      testGoldens(
          'screenMatchesGolden method can disable enableRealShadows temporarily',
          (tester) async {
        await tester.pumpWidgetBuilder(WidgetWithShadows());
        await screenMatchesGolden(
            tester, 'my_golden');
      })
    },
    config: GoldenToolkit.configuration.copyWith(enableRealShadows: false),
  );
});
```

fixes #72

### CODE REVIEW

File: packages/golden_toolkit/lib/src/testing_tools.dart [hunk 1]
```diff
@@ -151,9 +151,14 @@ void testGoldens(
...
         _inGoldenTest = true;
         tester.binding.addTime(const Duration(seconds: 10));
+        final initialDebugDisableShadowsValue = debugDisableShadows;
+        final shouldUseRealShadows =
+            GoldenToolkit.configuration.enableRealShadows;
+        debugDisableShadows = !shouldUseRealShadows;
         try {
           await test(tester);
         } finally {
+          debugDisableShadows = initialDebugDisableShadowsValue;
```
Comment (tsimbalar) [inline packages/golden_toolkit/lib/src/testing_tools.dart:161]: this is the actual change ... all the rest is mostly tests :P

File: packages/golden_toolkit/test/configuration_test.dart [hunk 2]
```diff
@@ -173,6 +175,40 @@ void main() {
...
 
+    group('Shadows', () {
+      GoldenToolkit.runWithConfiguration(
+        () {
+          testGoldens(
+              'screenMatchesGolden method uses enableRealShadows from global configuration when GoldenToolkit.runWithConfiguration is outside testGoldens',
+              (tester) async {
+            await tester.pumpWidgetBuilder(WidgetWithShadows());
+            await screenMatchesGolden(
+                tester, 'enableRealShadows_honored_when_testGoldens_wrapped');
```
Comment (tsimbalar) [inline packages/golden_toolkit/test/configuration_test.dart:186]: you'll see that `enableRealShadows_honored_when_testGoldens_wrapped.png` has nice shadows :)

File: packages/golden_toolkit/test/configuration_test.dart [hunk 3]
```diff
@@ -173,6 +175,40 @@ void main() {
...
+      );
+
+      testGoldens(
+          'screenMatchesGolden method does not use enableRealShadows from global configuration when GoldenToolkit.runWithConfiguration is inside testGoldens',
+          (tester) async {
+        await GoldenToolkit.runWithConfiguration(
+          () async {
+            await tester.pumpWidgetBuilder(WidgetWithShadows());
+            await screenMatchesGolden(tester,
+                'enableRealShadows_ignored_when_testGoldens_not_wrapped');
```
Comment (tsimbalar) [inline packages/golden_toolkit/test/configuration_test.dart:202]: you'll see that `enableRealShadows_ignored_when_testGoldens_not_wrapped.png` has ugly opaque block shadows

File: packages/golden_toolkit/test/sample_widgets.dart [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import 'package:flutter/material.dart';
+import 'package:flutter/widgets.dart';
+
+class WidgetWithShadows extends StatelessWidget {
```
Comment (tsimbalar) [inline packages/golden_toolkit/test/sample_widgets.dart:4]: just a widgets with drop shadows, useful to test that indeed shadows are drawn properly or not

File: packages/golden_toolkit/test/shadows/flutter_test_config.dart [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+      await testMain();
+    },
+    config: GoldenToolkitConfiguration(
+      // Currently, goldens are not generated/validated in CI for this repo. We have settled on the goldens for this package
+      // being captured/validated by developers running on MacOSX. We may revisit this in the future if there is a reason to invest
+      // in more sophistication
+      skipGoldenAssertion: () => !Platform.isMacOS,
+      // enableRealShadows can only work when "outside" of a `testGoldens` , for instance here at the root
+      enableRealShadows: true,
+    ),
```
Comment (tsimbalar) [inline packages/golden_toolkit/test/shadows/flutter_test_config.dart:27]: in the `shadows` folder, we set `enableRealShadows: true` , and all tests in this folder or subfolder will auto-magically have "nice shadows"

File: packages/golden_toolkit/test/shadows/shadows_test.dart [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    tearDown(() {
+      expect(debugDisableShadows, isTrue,
+          reason: 'debugDisableShadows should always be true after test');
+    });
+
+    testGoldens(
+        'screenMatchesGolden method uses enableRealShadows from global configuration (flutter_test_config.dart)',
+        (tester) async {
+      await tester.pumpWidgetBuilder(WidgetWithShadows());
+      await screenMatchesGolden(tester, 'realShadows_enabled_globally');
```
Comment (tsimbalar) [inline packages/golden_toolkit/test/shadows/shadows_test.dart:32]: here, without doing anything special in the test, we'll get real shadows in out goldens

File: packages/golden_toolkit/test/shadows/shadows_test.dart [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+            (tester) async {
+          await tester.pumpWidgetBuilder(WidgetWithShadows());
+          await screenMatchesGolden(tester, 'realShadows_disabled_locally');
+
+          expect(debugDisableShadows, isTrue,
+              reason: 'debugDisableShadows should be false during test');
+        })
+      },
+      config: GoldenToolkit.configuration.copyWith(enableRealShadows: false),
+    );
```
Comment (tsimbalar) [inline packages/golden_toolkit/test/shadows/shadows_test.dart:52]: an example of disabling shadows for a test in concrete when "nice shadows" have been enabled at a higher lever

**Comment by coreysprague** [comment]
This looks good to me. Thanks for the contribution @tsimbalar!

**Review by coreysprague — APPROVED** [review]
👍

**Pull request merged**